File: leetcode/problems/cpp/valid-parentheses.cpp
#include "stack"
#include "string"
#include "vector"
using namespace std;

// by @codeAntu
// time complexity: O(n)
// space complexity: O(n)

class Solution {
   inline bool match(char a, char b) {
      return (a == '(' && b == ')') || (a == '{' && b == '}') || (a == '[' && b == ']');
   }

  public:
   bool isValid(string s) {
      stack<char> stk;
      for (int i = 0; i < s.length(); i++) {
         if (s[i] == '(' || s[i] == '{' || s[i] == '[')
            stk.push(s[i]);
         else if (!stk.empty() && match(stk.top(), s[i]))
            stk.pop();
         else
            return false;
      }
      return stk.empty();
   }
};

File: leetcode/problems/rs/valid-parentheses.rs
// by @codeAbinash
// Time : O(n)
// Space : O(n)

pub struct Solution;

impl Solution {
   pub fn is_valid(s: String) -> bool {
      let mut stack = Vec::new();
      for c in s.chars() {
         match c {
            '(' | '[' | '{' => stack.push(c),
            ')' => {
               if stack.pop() != Some('(') {
                  return false;
               }
            }
            ']' => {
               if stack.pop() != Some('[') {
                  return false;
               }
            }
            '}' => {
               if stack.pop() != Some('{') {
                  return false;
               }
            }
            _ => unreachable!(),
         }
      }
      stack.is_empty()
   }
}
